emeval/input/spec_details.py

```diff
@@ -138,4 +114,71 @@ def get_geojson_for_leg(cls, gt_leg):
         else:
             gt_leg["loc"]["properties"]["style"] = {"color": "purple", "fillColor": "purple"}
             return gt_leg["loc"]
-        
+
+
+class ServerSpecDetails(SpecDetails):
+    def retrieve_one_batch(self, user, key_list, start_ts, end_ts):
+        post_body = {
+            "user": user,
+            "key_list": key_list,
+            "start_time": start_ts,
+            "end_time": end_ts
+        }
+
+        print(f"Retrieving data for: {post_body=}")
+        try:
+            response = requests.post(f"{self.DATASTORE_LOC}/datastreams/find_entries/timestamp", json=post_body)
+            print(f"{response=}")
+            response.raise_for_status()
+            data = response.json()["phone_data"]
+        except Exception as e:
+            print(f"Got {type(e).__name__}: {e}, retrying...")
+            time.sleep(10)
+            response = requests.post(f"{self.DATASTORE_LOC}/datastreams/find_entries/timestamp", json=post_body)
+            print(f"{response=}")
+            response.raise_for_status()
+            data = response.json()["phone_data"]
+
+        for e in data:
+            e["data"]["write_ts"] = e["metadata"]["write_ts"]
+
+        print(f"Found {len(data)} entries")
+        return data
+
+    def retrieve_data(self, user, key_list, start_ts, end_ts):
+        all_done = False
+        location_entries = []
+        curr_start_ts = start_ts
+        prev_retrieved_count = 0
+
+        while not all_done:
+            print("Retrieving data for %s from %s -> %s" % (user, curr_start_ts, end_ts))
+            curr_location_entries = self.retrieve_one_batch(user, key_list, curr_start_ts, end_ts)
+            #print("Retrieved %d entries with timestamps %s..." % (len(curr_location_entries), [cle["data"]["ts"] for cle in curr_location_entries[0:10]]))
+            if len(curr_location_entries) == 0 or len(curr_location_entries) == 1:
+                all_done = True
+            else:
+                location_entries.extend(curr_location_entries)
+                new_start_ts = curr_location_entries[-1]["metadata"]["write_ts"]
+                assert new_start_ts > curr_start_ts
+                curr_start_ts = new_start_ts
+                prev_retrieved_count = len(curr_location_entries)
+        return location_entries
+
+
+class FileSpecDetails(SpecDetails):
+    def retrieve_data(self, user, key_list, start_ts, end_ts):
+        data = []
+        for key in key_list:
+            data_file = os.path.join(
+                os.getcwd(),
+                self.DATASTORE_LOC,
+                f"{user}/{self.CURR_SPEC_ID}/{key.replace('/', '~')}/{math.floor(start_ts)}_{math.ceil(end_ts)}.json")
+            assert os.path.isfile(data_file), f"not found: {data_file=}"
+            with open(data_file, "r") as f:
+                d = json.load(f)
+                if isinstance(d, list):
+                    data.extend(d)
+                else:
+                    data.append(d)
```

Comment:
When do you get a dictionary instead of a list? Is that for the spec details?

Reply:
Yes this occurs with the spec details. This fix addressed all corner cases there.